src/crimson/os/seastore/extent_placement_manager.h

```diff
@@ -549,29 +698,146 @@ class ExtentPlacementManager {
       }
     }
 
-    bool background_should_run() const {
+    bool background_should_run() {
       assert(is_ready());
-      return cleaner->should_clean_space()
-        || trimmer->should_trim_dirty()
-        || trimmer->should_trim_alloc();
+      maybe_update_eviction_mode();
+      return main_cleaner_should_run()
+        || cold_cleaner_should_run()
+        || trimmer->should_trim();
+    }
+
+    bool main_cleaner_should_run() const {
+      assert(is_ready());
+      return main_cleaner->should_clean_space() ||
+        (has_cold_tier() &&
+         main_cleaner->can_clean_space() &&
+         eviction_state.is_fast_mode());
+    }
+
+    bool cold_cleaner_should_run() const {
+      assert(is_ready());
+      return has_cold_tier() &&
+        cold_cleaner->should_clean_space();
     }
 
     bool should_block_io() const {
       assert(is_ready());
       return trimmer->should_block_io_on_trim() ||
-             cleaner->should_block_io_on_clean();
+             main_cleaner->should_block_io_on_clean();
```

Comment:
Seems there is a missing piece to consider `cold_cleaner->should_block_io_on_clean()` properly with tiering?

Reply:
Write through policy is not implemented currently, so it doesn't cause errors without reservation on the cold cleaner.

So the only thing to consider is the semantics of reserving 0 on the cold cleaner when `cold_cleaner->should_block_io_on_clean()` returns true. I think the IO shouldn't be blocked in such cases. So the new implementation might be like this:

```cpp
bool SegmentCleaner::try_reserve_projected_usage(std::size_t projected_usage)
{
  assert(background_callback->is_ready());
  stats.projected_used_bytes += projected_usage;
  if (should_block_io_on_clean() && projected_usage != 0) { // <-- bypass the block condition
    stats.projected_used_bytes -= projected_usage;
    return false;
  } else {
    ++stats.projected_count;
    stats.projected_used_bytes_sum += stats.projected_used_bytes;
    return true;
  }
}
```

Reply:
IIUC there is another assumption of background cleaning is that it will become less efficient with the device(tier) become more full.

At a certain point (configurable) the clean effort will be treated unacceptable (the write amplification cost to store more data), and the cleaner needs to notify EPM that no more external I/O should be to this tier.

Reply:
Hm, another issue might also be related, that the unalive space doesn't mean that they are immediately available for write. So `should_block_io_on_clean()` is introduced to make sure the available space is always sufficient by blocking the incoming I/O.

Reply:
Seems we can use a new IO path in the future:

1. select generation for extent,
2. check whether the target generation is able to write,
  a. true, reserve the usage,
  b. false, return step 1 to select another tier's generation. When there is no more tier to select then the transaction is blocked on the last tier. 
3. write ool extents and submit the transaction.

Currently just add condition `cold_cleaner->should_block_io_on_clean()` to `EPM::should_block_io_on_clean()`

Reply:
Maybe my direction is incorrect.

Looking at the logic, `should_block_io_on_clean()` is already included in `try_reserve_projected_usage()`, and `should_block_io_on_trim()` included in `try_reserve_inline_usage()`.

Which means that the reservations already covers these conditions, so probably just drop `BackgroundProcess::should_block_io()`, and move `should_block_...()` to their private sections.

Reply:
Looks reasonable, `BackgroundProcess::should_block_io()` is only invoked by `BackgroundProcess::maybe_wake_blocked_io()`, which is not the accurate block condition that whether the current projected usage + IO usage exceeds the threshold.

I'll remove this method and make `AsyncCleaner::should_block_io_on_clean` and `JournalTrimmer::should_block_io_on_trim` private.

Reply:
@zhscn sorry for the backforth discussions.

I just realized there is a reason to keep it, which is to reduce the number to wake up the `blocking_io` promise.

The cleaners are working independently, so the BackgroundProcess here tries to query the background workers whether they all agree to unblock the io, when one of the worker becomes available.

So as an optimization (probably minor), I think its fine to `add condition cold_cleaner->should_block_io_on_clean() to EPM::should_block_io_on_clean()`, or `drop BackgroundProcess::should_block_io() entirely`.